Fixes #10644 Process>>terminate - fix failing test etc.

isCzech opened the pull request:
fix failing testIsTerminatingForcedTermination (currently as expected failure) and further deficiencies like correctly terminate non-local returns in ensure blocks, correctly terminate nested errors in ensure blocks and nested terminations, and allow to safely debug BlockCannotReturn scenarios.
E.g. the following examples will be evaluated correctly and safely:
[[[] ensure: [^2]] ensure: [^42]] fork
"any debugging possible unless pressed Proceed"
More scenarios in tests pushed separately soon.

This work extends and simplifies changes to #terminate merged  previously in #8993.

MarcusDenker committed:
This PR adds support (and a test) for metalinks on Pragmas. (Test was from #8374)

- add testSetLinkOnPragma
- adds support in the compiler (see #visitPragmaNode:)
- adds support to ask for reifications on RBPragmaNode

isCzech committed:
Fixes #10644 Process>>terminate - fix failing test, correctly terminate non-local returns in ensure blocks and nested terminations etc.

dionisiydk commented:
Did you forget to remove #testIsTerminatingForcedTermination from expected failures?

BTW there is a pragma <expectedFailure> to mark given test directly.

dionisiydk commented on src/Kernel/Context.class.st:
```diff
@@ -1772,6 +1772,31 @@ Context >> runUntilErrorOrReturnFrom: aSender [
+
+	Processor activeProcess
+		evaluate: [ 
+			| here unwindBottom newTop |
```
I would split this method on two parts: 
- Context>>runUnwindsTo: with this code inside #evaluate (or similar).
- And Process>>#runUnwindsBetween:to: which would wrap "aContext runUnwindsTo:" with #evaluate:onBehalfOf:. See other senders of #evaluate:onBehalfOf: . They are all on the Process side.

isCzech replied:
yes, I was considering this and checking typical senders of #evaluate:onBehalfOf: but then I decided against; no "big" reason though: Context it just seemed to me a better place for a method manipulating context similarly as #runUntilErrorOrReturnFrom: . #runUnwindTo:onBehalfOf: was conceived by pruning #runUntilErrorOrReturnFrom: as much as possible :) But you're right, it could as well be under Process...

isCzech replied:
one of the reasons I put #runUnwind:onBehalfOf: under Context is I can emphasize the fact it's evaluated "on behalf of" in the name of the method; and other could be the process being terminated is the patient being operated on, so making it a receiver (the one object "responsible" for the action in the OO philosophy) didn't seem right :)

dionisiydk commented on src/Kernel/Process.class.st:
```diff
@@ -213,63 +213,46 @@ Process >> debugWithTitle: title [
+		[ctxt isNil] whileFalse: [
+			ctxt unwindComplete ifNotNil: [
+				outerMost := ctxt].
+			ctxt := ctxt findNextUnwindContextUpTo: nil].
```
Let's introduce Context>>findMostOuterUnwindContext

isCzech replied:
I haven't made my mind about this yet. There might be some more changes here in the future and splitting the method into smaller ones makes it, for me at least, less readable. I need to see the whole code in this case :) (I don't do this generally but in this particular case with a tricky code like #teminate I prefer less fragmented code). When the code "settles" I'm definitely not against fragmenting. I just know how many times I reshuffled this code which is why I'm presenting it as an unfragmented method.

dionisiydk commented on src/Kernel/Process.class.st:
```diff
@@ -213,63 +213,46 @@ Process >> debugWithTitle: title [
+				newTop := ctxt unwindBlock asContextWithSender: ctxt.
+				newTop runUnwindTo: newTop onBehalfOf: self].
+			ctxt := ctxt findNextUnwindContextUpTo: nil].
+		suspendedContext := top setSender: nil receiver: self method: (Process>>#endProcess) arguments: {}
```
And I would extract it into Context>>beDead method

isCzech commented:
> Did you forget to remove #testIsTerminatingForcedTermination from expected failures?
> 
Yes, I did :)

isCzech committed:
remove #testIsTerminatingForcedTermination from expectedFailures

isCzech committed:
add some basic tests documenting:
- nested unwinds and semantics with non-local return
- process faithful termination - keeping terminating process identity
- behavior when a process terminating other process gets terminated before the termination is completed - the point being both processes will get properly unwound/terminated.

isCzech committed:
remove #expectedFailures method and replace it with <expectedFailure> pragma; however, because the method generates an error which fails the jenkins test I commented out the body of the method for lack of better ideas :)

dionisiydk commented on src/Kernel/Process.class.st:
```diff
@@ -213,63 +213,46 @@ Process >> debugWithTitle: title [
+		suspendedContext := oldList handleProcessTerminationOfWaitingContext: suspendedContext.
+
+		top := suspendedContext.
+		suspendedContext := nil.  "Disable terminating process while running its stack in active process below.
```
I am not sure about the idea of nil'ed suspendedContext. When debugger shows ZeroDivide in my example I can check the process in the playground: 
- it is terminated: p isTerminated "==> true"
- p suspendedContext "==> is nil"

This state is lying because the process is not yet terminated: it is in the middle of termination.  Why this nil is required? We have #terminating flag which protects from multiple termination.

isCzech replied:
Yes, #terminating could be enough... I don't think nil is "required" - I didn't use it for anything; it just seemed safer just nil it. suspendedContext of a terminating process has no longer any meaning during termination (except the last step) because it's not used during termination. But you're right it implicates #isTerminated value... I'll think about this. However, if there's no practical use for suspendedContext or isTerminated showing false during termination - I'd say it's more a philosophical question :)

dionisiydk replied:
> Yes, #terminating could be enough... I don't think nil is "required" - I didn't use it for anything; it just seemed safer just nil it. 

I prefer to keep the system simple. Any line of code raises the question "why it is here?". It's better to remove redundant code. Otherwise you should describe the need for this code with a test.

>> suspendedContext of a terminating process has no longer any meaning during termination (except the last step) because it's not used during termination. 

But it is used during termination. All unwinds are started from the suspendedContext. I think the unwind logic can better reflect the unwind steps: it can assign #newTop to #suspendedContext on each iteration. It will make the "outside termination" to be consistent with the "inside termination" (from the process itself).

>> However, if there's no practical use for suspendedContext or isTerminated showing false during termination - I'd say it's more a philosophical question :)

Having the system in the consistent state is very important. Especially when we implement such a very corner cases. It's always hard to debug and when in one place the process is shown as dead and in another place it is not dead that is a nightmare to figure things out.

isCzech replied:
Hi Denis,
yes, you have a point indeed and I hear your arguments! My objective now is to make sure it works flawlessly (and I have tested it quite thoroughly) before any further polishing. But I'm working on your comments, thanks again!

isCzech replied:
Hi, thanks for these great points!
> > Yes, #terminating could be enough... I don't think nil is "required" - I didn't use it for anything; it just seemed safer just nil it.
> 
> I prefer to keep the system simple. Any line of code raises the question "why it is here?". It's better to remove redundant code. Otherwise you should describe the need for this code with a test.
> 
Oh, I know why - if the suspendedContext is not nil the process being terminated can be resumed (e.g. by a higher priority process or thanks to preemption) and that would interfere with its termination in progress. Setting suspendedContext to nil prevents such scenario (see Process >> resume). In case this makes a good sense, I will update the comment; even I forgot the reasoning :D 

> > > suspendedContext of a terminating process has no longer any meaning during termination (except the last step) because it's not used during termination.
> 
> But it is used during termination. All unwinds are started from the suspendedContext. I think the unwind logic can better reflect the unwind steps: it can assign #newTop to #suspendedContext on each iteration. It will make the "outside termination" to be consistent with the "inside termination" (from the process itself).
> 
yes, right; I meant it has no reasonable meaning beyond the point the real unwinds start (i.e. after top := suspendedContext). The point is even if I set suspendedContext to newTop each time, its consistency is lost when the active process jumps onto the terminating process's stack and starts executing it's code. I remember when I first drafted this new approach I used suspendedContext the way you suggest but it kept failing for some reason so I gave up and left it as nil and used a variable instead (I'm not saying it's wrong; I just didn't manage to make it work that way).

> > > However, if there's no practical use for suspendedContext or isTerminated showing false during termination - I'd say it's more a philosophical question :)
> 
> Having the system in the consistent state is very important. Especially when we implement such a very corner cases. It's always hard to debug and when in one place the process is shown as dead and in another place it is not dead that is a nightmare to figure things out.
>
yes, I couldn't agree more; consistency is absolutely key; however in this case I don't know how to make isTerminated showing false during termination when I need suspendedContext = nil without further complicating isTerminated; actually maybe we could make isTerminated false if isTerminating = true -?? That should fix the consistency issue, what do you think? I'm not sure if all tests would agree with this change right away but it could be dealt with independently of this PR. Other option could be to change Process >> resume to take isTerminating into account and then we could remove nilling suspendedContext and leave isTerminated as is -?? (still independently of this PR)

dionisiydk replied:
> Oh, I know why - if the suspendedContext is not nil the process being terminated can be resumed (e.g. by a higher priority process or thanks to preemption) and that would interfere with its termination in progress. Setting suspendedContext to nil prevents such scenario (see Process >> resume). In case this makes a good sense, I will update the comment; even I forgot the reasoning :D

Right, I see that #resume method contains nil check for suspendedContext.
I think for the real solution we should explicitly model the process status - possibly replace #isTerminating flag with #status variable (choose a better name) and reify the ProcessStatus hierarchy. It will allow to decouple the life cycle state of a process from its execution state like suspendedContext, pc, endProcess method.

Anyway it's for future work. PR is already a great improvement.

isCzech replied:
Hi @dionisiydk , you wrote
> I think the unwind logic can better reflect the unwind steps: it can assign #newTop to #suspendedContext on each iteration.

FYI, you're right, I tried it and it works ok and makes isTerminated answer correctly; the only issue is you're not protected against accidental resumption of the terminating process.

dionisiydk reviewed (approved):
Good stuff

Alisu committed:
Add a roundedFractionPart for ScaledDecimal and use it for having coherent comparison (with tests)

Alisu committed:
Missing from previous commit

Alisu committed:
Hotfix c/c is bad

Alisu committed:
Update tests and fix properly

Alisu committed:
Also update the hash with tests

Alisu committed:
Adding comment

Alisu committed:
More readable code

Alisu committed:
Fixing failing tests

Alisu committed:
Remove part of the test code that was testing more the internal representation than the behavior.

Alisu committed:
See last commit

johnhwoods committed:
Fix for issue 11070 with test

MarcusDenker committed:
Merge pull request #11097 from pharo-project/Pharo10

Merging latest changes in P10 to P11

MarcusDenker committed:
Merge pull request #11086 from johnhwoods/11070-Interval-and-Intervalhash-do-not-match

Fix for issue 11070: Interval>>= and Interval>>hash do not match

RemiDufloer committed:
Fix #11109 
Move startSteppingSubmorphs to Morph, replace senders, remove old method

RemiDufloer committed:
Fix for the issue #11115, we assign  the pathTransform in the currentClip

MarcusDenker committed:
add Pharo11 to #platformAttributes

MarcusDenker committed:
add accessor to AbstractInitializedSlot

fixes #11080

MarcusDenker committed:
Merge pull request #11118 from MarcusDenker/11099-Pharo11-build-broken-Error-Could-not-determine-image-version-Using-fallback

11099-Pharo11-build-broken-Error-Could-not-determine-image-version-Using-fallback

Ducasse committed:
Merge pull request #11110 from RemiDufloer/issues/11109

Fix #11109: Feature Envy on startSteppingSubmorphsOf:

estebanlm committed:
add pharo11.x to metacelloPlatformAttributes

estebanlm committed:
use development versions

MarcusDenker committed:
Merge pull request #11120 from MarcusDenker/11080-InitializedSlot-did-not-understand-default

add accessor to AbstractInitializedSlot

MarcusDenker committed:
testProperAccessingProtocolIsUsed can now check for 0

MarcusDenker committed:
fix comments mentioning compact classes.

fixes #11127

MarcusDenker committed:
Merge pull request #11128 from MarcusDenker/testProperAccessingProtocolIsUsed-Zero

testProperAccessingProtocolIsUsed can now check for 0

MarcusDenker committed:
This PR Refactors the Protocol hierachy. By not inheriting vrom Protocol, AllProtocol does not need to allocate space for the two ivars that it never uses (name and methodSelectors).

(there are numberOfClasses*2 instances of AllProtocol)

fixes #10963

MarcusDenker committed:
Merge pull request #11132 from MarcusDenker/11127-Update-comment-in-ObjectprimitiveChangeClassTo-to-not-discuss-compact-classes

fix comments mentioning compact classes.

MarcusDenker committed:
Fix ExternalChangesBrowser as described in the issue tracker entry: use #selectedItemsSorted

fixes 11063

plantec committed:
fixes #11159

enrich KeyboardKey with key accessing methods

Ducasse committed:
Merge pull request #11160 from plantec/11159-Complete-the-set-of-methods-to-access-keys

11159-Complete-the-set-of-methods-to-access-keys

MarcusDenker committed:
Merge pull request #11156 from MarcusDenker/11063-ExternalChangesBrowserbuildMenu-should-use-selectedItemsSorted

11063-ExternalChangesBrowserbuildMenu-should-use-selectedItemsSorted

MarcusDenker committed:
fixes 11162 by removing ReToDoRule and tests

MarcusDenker committed:
improve comment of newFrom:

fixes #6719

MarcusDenker committed:
fix typo

MarcusDenker committed:
This PR moves the creation of the empty OrderedCollection from #sharedPools to #addSharedPool:.

This avoids allocating an OrderedCollection when e.g. creating AST (which happens on comping a method or looking at it in the browser).

This PR will not make the boostrapped image smaller, but it saves memory later (e.g. for newly loaded classes)

fixes #10957

MarcusDenker committed:
fix typo

Ducasse committed:
Merge pull request #11171 from MarcusDenker/10957-Memory-sharedPools-OrderedCollections-waste-memory

10957-Memory-sharedPools-OrderedCollections-waste-memory

MarcusDenker committed:
Merge pull request #11166 from MarcusDenker/6719-Dictionary-creation-API-should-accept-associations-too

6719-Dictionary-creation-API-should-accept-associations-too

MarcusDenker committed:
fix argument name

MarcusDenker committed:
Merge pull request #11148 from MarcusDenker/10963-Memory-AllProtocol-uses-a-lot-of-memory

10963-Memory-AllProtocol-uses-a-lot-of-memory

kasperosterbye committed:
Inserted check for empty prefix/suffix

kasperosterbye committed:
Changed test to test new definitions

kasperosterbye committed:
a few places simplifies with new difinition

hogoww committed:
[ci skip] [cleanup] improve comment of directories and files methods

MarcusDenker committed:
Merge pull request #11174 from hogoww/fixComment

[ci skip] [cleanup] improve comment of directories and files methods

MarcusDenker committed:
push intialization to basicDeclareClassVariable:. This way we only create those dictioaries that are needed. This should save some 600Kb of memory for the current Pharo11 image.

fixes #11172

MarcusDenker committed:
Change #newFrom: to do the same as #newFromAssociations: Instead of error, "If the same key appears twice in anAssocCollection, the last one enumerated will win"

fixes #6720

jordanmontt committed:
Deleted the line that was updateing the color of the label of the button presenter. Now, in Spec2, we have to use spec styles for the colors and so on.

jordanmontt committed:
Added shortcut method to show the formatting key bindings.

Ducasse committed:
Merge pull request #11181 from jordanmontt/11108-Add-shortcut-description-to-code-format-menu

Added shortcut method to show the formatting key bindings.

MarcusDenker committed:
Merge pull request #11180 from jordanmontt/11143-MiniDrTests-raise-error

Fixes issue #11143: MiniDrTests raise error

MarcusDenker committed:
Rename BlockClosure startpc to compiledBlock

fixes #11177

MarcusDenker committed:
Merge pull request #11175 from MarcusDenker/11172-ClassclassPool-empty-dictionary-for-all-classes-but-not-many-define-class-vars

11172-ClassclassPool-empty-dictionary-for-all-classes-but-not-many-define-class-vars

MarcusDenker committed:
Merge pull request #11188 from pharo-project/BlockClosure-rename-startPC

Rename BlockClosure startpc to compiledBlock

MarcusDenker committed:
remove unused LocaleAnnouncementTest, fixes #11144

MarcusDenker committed:
Merge pull request #10436 from MarcusDenker/Reflectivity-AddPragmaSupport

Reflectivity-AddPragmaSupport

MarcusDenker committed:
Start to remove code from Deprecated10 package. We do it in steps to make sure we do not break anything

kasperosterbye committed:
Merge branch 'pharo-project:Pharo11' into 11165-beginsWith-and-endsWith-should-return-true-on-empty-string

MarcusDenker committed:
Merge pull request #11192 from MarcusDenker/Cleanup-Deprecated10-1

Start to remove code from Deprecated10 package.

MarcusDenker committed:
I propose to not raise an error for the trait composition for duplications. 

- we should accept it
- then we can in the tools guide people to fix it

MarcusDenker committed:
Merge pull request #11189 from MarcusDenker/11144-LocaleAnnouncementTest-seems-to-be-unused

11144-LocaleAnnouncementTest-seems-to-be-unused

MarcusDenker committed:
Guard adding the temp for return value. #11198 for Pharo11

MarcusDenker committed:
Merge pull request #11199 from MarcusDenker/11198-Pharo11-Reflectivity-do-not-add-temp-for-value-if-not-generating-the-primtive-Wrapper

11198-Pharo11-Reflectivity-do-not-add-temp-for-value-if-not-generating-the-primtive-Wrapper

MarcusDenker committed:
fixes #6323 as described in the issue tracker entry

MarcusDenker committed:
Merge pull request #11117 from RemiDufloer/issues/11115

Fix for the issue #11115: Feature Envy clipRect on AthensCairoCanvas

hellerve committed:
Implement EpInverseVisitor>>#visitTraitModification: [pharo-project/pharo#11157]

MarcusDenker committed:
Two tests fixed, one step for #11195

MarcusDenker committed:
Merge pull request #11208 from hellerve/Pharo11

Implement EpInverseVisitor>>#visitTraitModification:

MarcusDenker committed:
Merge pull request #11173 from kasperosterbye/11165-beginsWith-and-endsWith-should-return-true-on-empty-string

Fixes #11165 begins with and ends with should return true on empty string

MarcusDenker committed:
Merge pull request #11209 from MarcusDenker/Fix-messagetally-tests-Cleanblocks

Fix-messagetally-tests-Cleanblocks

MarcusDenker committed:
Merge pull request #10951 from Alisu/Pharo10

FIX issue #10933: Scaled decimals comparison result depends on the number creation method

MarcusDenker committed:
Merge pull request #11205 from MarcusDenker/6323-RubSmalltalkEditor-meta-j-shortcut-not-working

6323-RubSmalltalkEditor-meta-j-shortcut-not-working

demarey committed:
Fix an issue where streaming client usage failed when the server set a connection close.

+ sync of zinc repository

MarcusDenker committed:
Merge pull request #11164 from MarcusDenker/11162-Check-ReToDoRule

11162-Check-ReToDoRule

MarcusDenker committed:
Merge pull request #11176 from MarcusDenker/6720-Dictionary--newFrom-logic-can-be-improved-

6720-Dictionary--newFrom-logic-can-be-improved-

MarcusDenker committed:
Small cleanup, fixes #10185

MarcusDenker committed:
fixes #11216

MarcusDenker committed:
- make CleanBlockClosure a subclass of BlockClosure
- BlockClosure is now tagged abstract
- override #receiver  in CleanBlockClosure
- push down #sender
- Improve comments

fixes #11217

MarcusDenker committed:
add back ivar for the VM

MarcusDenker committed:
add accessor

MarcusDenker committed:
Merge pull request #11224 from MarcusDenker/11216-categorize-ZnStaticFileDecoratorDelegate-classdecorateservingFilesFrom

11216-categorize-ZnStaticFileDecoratorDelegate-classdecorateservingFilesFrom

MarcusDenker committed:
Merge pull request #11215 from demarey/fix/zinc-connection-close-11

 Fix an issue where streaming client usage failed when the server set a connection close.

MarcusDenker committed:
Merge pull request #11207 from pharo-project/Pharo10

Porting fixes from Pharo 10 to Pharo 11

MarcusDenker committed:
Revert "FIX issue #10933: Scaled decimals comparison result depends on the number creation method"

MarcusDenker committed:
The Compiler adds a  back pointer from compiled block to outer code.

Only do that for blocks the compiler created, where the outerCode is not yet set.

fixes #11218

MarcusDenker committed:
Merge pull request #11228 from pharo-project/revert-10951-Pharo10

Revert "FIX issue #10933: Scaled decimals comparison result depends on the number creation method"

MarcusDenker committed:
fixes 11230

CDSharedVariableNode is a subclass of CDSlotNode. This is not really good, as SharedVariables (ClassVariables) are not Slots.

The hierarchy should have an abstract CDVariableNode with the two classes being subclasses of that.

MarcusDenker committed:
This PR makes sure to nil out the SelectorTable on #cleanUp and #rehash.

This means that the download image will be a bit smaller (WeakSet of size 140K entries).

MarcusDenker committed:
do not record the #isDoIt selectors

MarcusDenker committed:
This commit changes the encoding as described in the Issue.

The MC reader and installer already assume utf8 and fall back to latin1 on error... we could do the same if needed
(see MCMczReader>>#contentStreamForMember: )

MarcusDenker committed:
Merge pull request #11229 from MarcusDenker/11218-CleanBlocks-Compiler-should-not-change-the-outerCode-for-all-blocks-in-the-literals

11218-CleanBlocks-Compiler-should-not-change-the-outerCode-for-all-blocks-in-the-literals

MarcusDenker committed:
Merge pull request #11225 from MarcusDenker/CleanBlock-Hierachy-Cleanup

CleanBlock-Hierachy-Cleanup

MarcusDenker committed:
Merge pull request #11231 from MarcusDenker/11230-ClassParser-Improve-Hierarchy-CDSharedVariableNodeCDSlotNode

11230-ClassParser-Improve-Hierarchy-CDSharedVariableNodeCDSlotNode

MarcusDenker committed:
Transcript output

MarcusDenker committed:
guard for boostrap

MarcusDenker committed:
more trying to get a log...

MarcusDenker committed:
more guards

MarcusDenker committed:
more...
- do not intern #UndefinedMethod
- log interning

MarcusDenker committed:
- Do not intern the selector when setting the selector on the compiledMethod, do it intead on #addSelectorSilently:withMethod:

This should avoid FFI generation to interfere.

MarcusDenker committed:
- remove logging
- simplify #internSelector:

MarcusDenker committed:
This PR removes changes to not initialize OneCharacterSymbols. 

- OneCharacterSymbols is not used by #intern:, just allSymbols
- Why add a symbo for all characters anyway? We will intern them as needed

The class variable is not yet removed.

This removed the unused method #selectorsContaining:, too

MarcusDenker committed:
Merge pull request #11235 from MarcusDenker/Reset-SelectorTable-onCleanup

Reset SelectorTable on #cleanUp

MarcusDenker committed:
add Symbol to testNoUnusedClassVariablesLeft (we remove the var later)

MarcusDenker committed:
Character>>asSymbol can just call #asSymbol on itself as a String

MarcusDenker committed:
Merge pull request #11241 from MarcusDenker/Cleanup-OneCharacterSymbols

Cleanup: OneCharacterSymbols: Symbol allSymbols contains one-character symbols twice

MarcusDenker committed:
Merge pull request #11219 from MarcusDenker/10185-merge-ReSmalllintChecker-into-SmalllintManifestChecker

10185-merge-ReSmalllintChecker-into-SmalllintManifestChecker

MarcusDenker committed:
remove OneCharacterSymbols from Symbol

This PR had to be done by changing the source directly

MarcusDenker committed:
Merge pull request #11247 from pharo-project/MarcusDenker-OneCharacter-Cleanup-Var

remove OneCharacterSymbols from Symbol

MarcusDenker committed:
- remove some methids tha are duplicate in the Hieachy
- simolify a method

aboubacardiawara committed:
Changing isLiteral categori to testing
issue: #11238

Commit:
Fixed issue 11210

aboubacardiawara committed:
Reving browserEnvironment method from TestRunner (any sender and it uses class which no longer exists).

MarcusDenker committed:
Merge pull request #11250 from MarcusDenker/Cleanup-DeadCode-05-20

Cleanup-DeadCode-05-20

MarcusDenker committed:
Merge pull request #11260 from bloomabir/Pharo11

Fixed issue 11210: Do not skip NativeArrayTest>>#testDoubleArraysHaveCorrectLayout

Ducasse committed:
Merge pull request #11239 from MarcusDenker/Pharo11-UFT8-Monticello

Pharo11-UFT8-Monticello

MarcusDenker committed:
Merge pull request #11257 from waradia09/Pharo11

[clean up] isLiteral and TestRunner method never call

isCzech committed:
After many iterations I've arrived at a code with a very similar structure as the original elegant code in Squeak versions 1 through 3.5, i.e. before Pharo forked :) I admired the simplicity of the original code using Context>>#unwindTo: to deal with all unwinds. The enclosed version of unwindTo: fixes issues with non-local returns and extends the unwind semantics a bit (see comments in the relevant methods).

The main highlights of the solution are:
1. direct termination - the unwind blocks are always executed by the process being terminated (i.e. termination is process faithful)
2. code structure follows (or copies) the structure used in the early Squeak implementation (versions 1 up to 3.5)
3. non-local returns can be placed anywhere inside unwind blocks, even nested etc.

Compare the proposed and the original #teminate from 1999 :)

isCzech committed:
Merge branch 'pharo-project:Pharo11' into 10644-Process-terminate---fix-failing-test-correctly-terminate-non-local-returns-in-ensure-blocks-and-nested-terminations-etc

Ducasse commented:
Hi jaromir
Thanks a lot! Now pay attention because you have a lot of changes in your commit. 
May be you should merge first P11 in your branch to realign it. 

S

isCzech commented:
Hi Stephane,
My apologies for the mess… I don’t know what happened, only wanted to push one commit (512bc5d) but when my repo synchronized it pushed 118 commits, never happened before but it’s surely my mistake :(
Thanks; have a nice day,
Jaromir

--

Jaromír Matas

***@***.***

From: ***@***.***>
Sent: Sunday, May 22, 2022 20:40
To: ***@***.***>
Cc: Jaromir ***@***.***>; ***@***.***>
Subject: Re: [pharo-project/pharo] Fixes #10644 Process>>terminate - fix failing test etc. (PR #10645)


Hi jaromir
Thanks a lot! Now pay attention because you have a lot of changes in your commit.
May be you should merge first P11 in your branch to realign it.

S

—
Reply to this email directly, view it on GitHub<https://github.com/pharo-project/pharo/pull/10645#issuecomment-1133951430>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ARPWHOKG53RLOOKKEOVNMO3VLJ5RLANCNFSM5J4JQVGA>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

Ducasse commented:
Do not worry :)
Probably your version was far and got confused.

MarcusDenker commented:
replaced by https://github.com/pharo-project/pharo/pull/11308, I close this
